packages/@lwc/engine-core/src/framework/hydration-utils.ts

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2024, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { ArrayPush, isNull, ArrayJoin, ArraySort, ArrayFrom } from '@lwc/shared';
+
+import { assertNotProd } from './utils';
+
+// Errors that occured during the hydration process
+let hydrationErrors: Array<HydrationError> = [];
+
+// These values are the ones from Node.nodeType (https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType)
+const enum EnvNodeTypes {
+    ELEMENT = 1,
+    TEXT = 3,
+    COMMENT = 8,
+}
+
+interface HydrationError {
+    type: string;
+    serverRendered: any;
+    clientExpected: any;
+}
+
+export type Classes = Omit<Set<string>, 'add'>;
+
+/*
+    Prints attributes as null or "value"
+*/
+export function prettyPrintAttribute(attribute: string, value: any): string {
+    assertNotProd(); // this method should never leak to prod
+    return `${attribute}=${isNull(attribute) ? attribute : `"${value}"`}`;
+}
+
+/*
+    Sorts and stringifies classes
+*/
+export function prettyPrintClasses(classes: Classes) {
+    assertNotProd(); // this method should never leak to prod
+    const value = JSON.stringify(ArrayJoin.call(ArraySort.call(ArrayFrom(classes)), ' '));
+    return `class=${value}`;
+}
+
+/*
+    Hydration errors occur before the source node has been fully hydrated,
+    queue them so they can be logged later against the mounted node.
+*/
+export function queueHydrationError(type: string, serverRendered?: any, clientExpected?: any) {
+    assertNotProd(); // this method should never leak to prod
+    ArrayPush.call(hydrationErrors, { type, serverRendered, clientExpected });
+}
+
+/*
+    Flushes (logs) any queued errors after the source node has been mounted.
+ */
+export function flushHydrationErrors(source?: Node | null) {
+    assertNotProd(); // this method should never leak to prod
+    for (let i = 0; i < hydrationErrors.length; i++) {
```

Comment:
No biggie, but for future reference `for...of` is fine. I think the old-school `for` loop still wins on micro-benchmarks, but it's not important for dev-only code.